src/harness/testcases/heartbeat_testcase.py

```diff
@@ -400,3 +401,100 @@ def test_10_9_4_2_3_1_3(self):
     response = self._sas.Heartbeat(request)['heartbeatResponse'][0]
     # Check the heartbeat response
     self.assertEqual(response['response']['responseCode'], 102)
+
+  @winnforum_testcase
+  def test_WINNF_FT_S_HBT_16(self):
+    """CBSD heartbeat request after grant is terminated.
+
+    Heartbeat request immediately after CBSD moves out of Granted State. The
+    grantId is invalid in heartbeat request. The response should be FAIL.
+    """
+
+    # Register the device
+    device_a = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_a.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    request = {'registrationRequest': [device_a]}
+    response = self._sas.Registration(request)['registrationResponse'][0]
+    # Check registration response
+    self.assertEqual(response['response']['responseCode'], 0)
+    cbsd_id = response['cbsdId']
+    del request, response
+
+    # Request grant
+    grant_0 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_0['cbsdId'] = cbsd_id
+    request = {'grantRequest': [grant_0]}
+    # Check grant response
+    response = self._sas.Grant(request)['grantResponse'][0]
+    self.assertEqual(response['cbsdId'], cbsd_id)
+    self.assertEqual(response['response']['responseCode'], 0)
+    grant_id = response['grantId']
+    grant_expire_time = datetime.strptime(response['grantExpireTime'],
+                                          '%Y-%m-%dT%H:%M:%SZ')
+    del request, response
+
+    # First successful Heartbeat
+    request = {
+        'heartbeatRequest': [{
+            'cbsdId': cbsd_id,
+            'grantId': grant_id,
+            'operationState': 'GRANTED'
+        }]
+    }
+    response = self._sas.Heartbeat(request)['heartbeatResponse'][0]
+    # Check the heartbeat response
+    self.assertEqual(response['cbsdId'], cbsd_id)
+    self.assertEqual(response['grantId'], grant_id)
+    self.assertLess(datetime.utcnow(),
+                    datetime.strptime(response['transmitExpireTime'],
+                                      '%Y-%m-%dT%H:%M:%SZ'))
+    self.assertEqual(response['response']['responseCode'], 0)
+
+    # Enter the Authorized state
+    request['heartbeatRequest'][0]['operationState'] = 'AUTHORIZED'
+    response = self._sas.Heartbeat(request)['heartbeatResponse'][0]
+    # Verify successful response
+    transmit_expire_time = datetime.strptime(response['transmitExpireTime'],
+                                             '%Y-%m-%dT%H:%M:%SZ')
+    self.assertEqual(response['cbsdId'], cbsd_id)
+    self.assertEqual(response['grantId'], grant_id)
+    self.assertLess(datetime.utcnow(), transmit_expire_time)
+    self.assertEqual(response['response']['responseCode'], 0)
+    del request, response
+
+    # Relinquish the grant
+    # Note: The Testcase requires grant to be terminated at this point, but
```

Comment:
We discussed with Yi Hsuan (one of the Google members of WG4) and he agreed that the test case should be changed since the more complex behavior specified in the test spec is not required to achieve the desired effect. We have not formally discussed with WG4 or test spec authors but plan to submit a comment when the spec is under ballot.

I see three options:
1) Submit the current code (knowing that we'll need to look carefully through all test cases once the spec is finalized anyway, and we can update at that time if needed)
2) Wait to submit until we get more formal agreement from WG4 (pre-ballot)
3) Write the code according to the test spec

(3) requires a lot more work for no gain; I expect we'll be successful in changing the test spec so I'd rather avoid (3). (2) isn't a bad option, especially if we think they will need to change other tests in a similar way. (1) also seems fine to me. Which do you prefer?